### PR TITLE
Add RoPE with FlexBert blocks

### PR DESCRIPTION
**Pull request by staghado**
This implements RoPE for BERT. See #3

It uses the rotary embedding kernel from [flash_attn](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/layers/rotary.py). In the unpadded version, we first have to pad -> rope -> unpad, this is cheap especially for the model sizes we are considering.

The unpadded rope version still achieves better throughput than the padded one at large sequence lengths even with the pad/unpad logic which means that the rotary operation is negligible compared to the attention.
compare-seqlen-rope

### CODE REVIEW

**Review by ohallstrom — CHANGES_REQUESTED** [review]
Looks good in general to me, but please look at the comments I have made and make the corresponding fixes.

Would appreciate if we could have another set of eyes on this PR as well, might missed something as its late..:)

**Comment by ohallstrom** [comment]
Also, a question to the other reviewers: Would we want to support of RoPE based on position IDs?

**Comment by ohallstrom** [comment]
Also, it seems like @staghado has incorporated some code from #35 , we should make sure these changes align fully with #35 once it has been merged

File: src/bert_layers/attention.py [hunk 1]
```diff
@@ -399,16 +403,234 @@ def forward(
...
+        self.p_dropout = config.attention_probs_dropout_prob
+        self.Wqkv = nn.Linear(config.hidden_size, 3 * self.all_head_size, bias=config.attn_qkv_bias)
+        self.Wo = nn.Linear(config.hidden_size, config.hidden_size, bias=config.attn_out_bias)
+        self.out_drop = (
+            nn.Dropout(config.attn_out_dropout_prob) if config.attn_out_dropout_prob > 0.0 else nn.Identity()
+        )
+
+        assert RotaryEmbedding is not None, "rotary_emb is not installed"
+        self.rotary_emb = RotaryEmbedding(
+            config.rotary_emb_dim,
```
[the remark this answers is not in the document]
Comment (iacolippo) [reply src/bert_layers/attention.py:443]: It depends if we want to enable partial rotary like in GPT-J. Not sure about it

File: src/bert_layers/attention.py [hunk 2]
```diff
@@ -399,16 +403,234 @@ def forward(
...
+            qkv = bert_padding.pad_input(qkv, indices, cu_seqlens.shape[0] - 1, max_seqlen)  # batch, max_seqlen, thd
+            unpad_bs, *_ = qkv.shape
+
+            # Reshape to (batch, seqlen, 3, nheads, headdim)
+            qkv = qkv.view(unpad_bs, -1, 3, self.num_attention_heads, self.attention_head_size)
+
+            # Apply RoPE
+            qkv = self.rotary_emb(qkv, seqlen_offset=seqlen_offset, max_seqlen=None)
+
+            qkv = qkv.view(unpad_bs, -1, 3, self.num_attention_heads, self.attn_head_size)
```
[the remark this answers is not in the document]
Comment (staghado) [reply src/bert_layers/attention.py:531]: yeah it's left from debugging

File: src/bert_layers/embeddings.py [hunk 3]
```diff
@@ -149,7 +149,7 @@ def __init__(self, config: FlexBertConfig):
         self.norm = get_norm_layer(config) if config.embed_norm else nn.Identity()
         self.drop = nn.Dropout(config.embed_dropout_prob) if config.embed_dropout_prob > 0.0 else nn.Identity()
 
-    def forward(self, input_ids: torch.LongTensor, **kwargs) -> torch.Tensor:
+    def forward(self, input_ids: torch.LongTensor, position_ids: Optional[torch.LongTensor] = None) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (staghado) [reply src/bert_layers/embeddings.py:152]: It's not used but is added in order to have an identical interface between **SansPositionEmbeddings**  and **AbsoluteEmbeddings**. Avoids having custom logic in the model forward.

File: yamls/main/flex-bert-base.yaml
[the remark this answers is not in the document]
Comment (staghado) [reply yamls/main/flex-bert-base.yaml:1]: I think it should as there is an inconsistent use of lower and uppercase for true/false. For train_small i'll remove it but it shouldn't be an issue as we will have our custom data splits.

File: src/bert_layers/attention.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ohallstrom) [reply src/bert_layers/attention.py:443]: If we want to enable partial rotary a solution is then simply just to set rotary_embed_size to None as default in https://github.com/AnswerDotAI/bert24/pull/36#discussion_r1615374459 . If None, then the embed dim should be set to attn_head_size. Else we can override it when we want to do partial rotary. That way we don't have to change the config twice if not doing partial rotary.

File: yamls/main/flex-bert-base.yaml
[the remark this answers is not in the document]
Comment (ohallstrom) [reply yamls/main/flex-bert-base.yaml:1]: In that case one could argue that it should be in a different PR as it has nothing to do with RoPE. But for me it is ok

File: src/bert_layers/attention.py [hunk 4]
```diff
@@ -404,16 +408,240 @@ def forward(
...
+            else:
+                attn = flash_attn_qkvpacked_func(qkv, dropout_p=self.p_dropout)
+        else:
+            qkv = qkv.view(batch_size, seqlen, 3, self.num_attention_heads, self.attn_head_size)
+            qkv = self.rotary_emb(qkv, seqlen_offset=seqlen_offset, max_seqlen=None)
+            q, k, v = qkv.transpose(3, 1).unbind(dim=2)
+            attn = F.scaled_dot_product_attention(q, k, v, dropout_p=self.p_dropout)
+            
+        # attn = attn.transpose(1, 2).view(batch_size, seqlen, dim).contiguous()
+        attn = attn.contiguous().view(batch_size, seqlen, dim)
```
Comment (ohallstrom) [inline src/bert_layers/attention.py:636]: missing a transpose(1, 2) here? attn has shape (bs, n_heads, seq_len, head_dim)

File: src/bert_layers/attention.py [hunk 5]
```diff
@@ -404,16 +408,240 @@ def forward(
...
+            qkv = qkv.view(unpad_bs, -1, 3, self.num_attention_heads, self.attn_head_size)
+
+            # Apply RoPE
+            qkv = self.rotary_emb(qkv, seqlen_offset=seqlen_offset, max_seqlen=None)
+
             q, k, v = qkv.transpose(3, 1).unbind(dim=2)  # b h s d
             attn = F.scaled_dot_product_attention(q, k, v, dropout_p=self.p_dropout).transpose(1, 2)
 
-        attn = attn.view(batch_size, seqlen, dim)
+            attn = attn.transpose(1, 2).view(unpad_bs, -1, dim)
```
Comment (ohallstrom) [inline src/bert_layers/attention.py:543]: double transpose(1, 2), one should be removed 😅

File: src/bert_layers/attention.py [hunk 6]
```diff
@@ -404,7 +408,226 @@ def forward(
                 attn = flash_attn_qkvpacked_func(qkv, dropout_p=self.p_dropout)
         else:
             qkv = qkv.view(batch_size, seqlen, 3, self.num_attention_heads, self.attn_head_size)
+
+            q, k, v = qkv.transpose(3, 1).unbind(dim=2)
+            attn = F.scaled_dot_product_attention(q, k, v, dropout_p=self.p_dropout)
+
+        attn = attn.contiguous().view(batch_size, seqlen, dim)
```
Comment (ohallstrom) [inline src/bert_layers/attention.py:415]: Itranspose(1,2) missingno?

File: src/bert_layers/attention.py [hunk 7]
```diff
@@ -404,7 +408,226 @@ def forward(
...
+                # bfloat16 must be supported as of FA2 2.5.7. (Turing GPUs not supported)
+                orig_dtype = qkv.dtype
+                qkv = qkv.to(torch.bfloat16)
+
+                attn = flash_attn_qkvpacked_func(qkv, dropout_p=self.p_dropout)
+                attn = attn.to(orig_dtype)  # type: ignore
+            else:
+                attn = flash_attn_qkvpacked_func(qkv, dropout_p=self.p_dropout)
+        else:
+            qkv = qkv.view(batch_size, seqlen, 3, self.num_attention_heads, self.attn_head_size)
```
Comment (ohallstrom) [inline src/bert_layers/attention.py:628]: This row (and 611) could be moved out of the if and else statement as both if and else start with the same row

**Review by ohallstrom — COMMENTED** [review]
LGTM overall, just address the two comments and I believe we can merge
